### PR TITLE
New braces_for_chaining style for Style/BlockDelimiters.

### PR DESCRIPTION
**Pull request by panthomakos**
As per [the style-guide][style-guide], even though multi-line chaining
is considered ugly and we should generally find ways to re-write the
code, sometimes this simply isn't possible. Regardless, some code-bases
would like to allow multi-line blocks to be used in method chains.

We can also argue that multi-line chaining, when used, is better like
this:

```
array.map { |x|
  x * 2
}.map(&:to_s)
```

than like this:

```
array.map do |x|
  x * 2
end.map(&:to_s)
```

This style, when selected, forces multi-line blocks to use braces if
their return value is being chained with another method. I believe
that this is a better option than the current default configuration
which enforces the second of the two code examples.

[style-guide]: https://github.com/bbatsov/ruby-style-guide

### CODE REVIEW

**Comment by jonas054** [comment]
This is already taken care of if you configure
```yaml
Style/BlockDelimiters:
  EnforcedStyle: semantic
```

**Comment by panthomakos** [comment]
@jonas054 This is a little bit different because `semantic` requires `{}` on assignment as well:

```
Style/BlockDelimiters:
  EnforcedStyle: semantic

# good
variable = array.map { |i|
  i*2
}
```

This PR allows you to enforce `do-end` for multi-line blocks except when chaining, so the previous example would be:

```
Style/BlockDelimiters:
  EnforcedStyle: line_count_based
  EnforceMultiLineBraceChaining: true

# good
variable = array.map do |i|
  i * 2
end

# good
array.map{ |i|
  i * 2
}.map(&:to_sym)
```

File: CHANGELOG.md [hunk 1]
```diff
@@ -17,6 +17,7 @@
 * [#2277](https://github.com/bbatsov/rubocop/pull/2277): New cop `Style/FirstHashElementLineBreak` checks for a line break before the first element in a multi-line hash. ([@panthomakos][])
 * [#2277](https://github.com/bbatsov/rubocop/pull/2277): New cop `Style/FirstMethodArgumentLineBreak` checks for a line break before the first argument in a multi-line method call. ([@panthomakos][])
 * [#2277](https://github.com/bbatsov/rubocop/pull/2277): New cop `Style/FirstMethodParameterLineBreak` checks for a line break before the first parameter in a multi-line method parameter definition. ([@panthomakos][])
+* [#2329](https://github.com/bbatsov/rubocop/pull/2329): New option `EnforceMultiLineBraceChaining` for `Style/BlockDelimiters` cop enforces braces on a multi-line block if it's return value is being chained with another method. ([@panthomakos][])
```
Comment (jonas054) [inline CHANGELOG.md]: it's -> its

**Comment by jonas054** [comment]
I see. Then I think it's OK. But since the new parameter only makes sense for the `line_count_based` style, I think it's better to implement it as a third style, `braces_for_chaining`.

**Comment by panthomakos** [comment]
@jonas054 PTAL I've updated with your comments and implemented using the `braces_for_chaining` style.

File: lib/rubocop/cop/style/block_delimiters.rb [hunk 2]
```diff
@@ -28,7 +28,7 @@ def on_block(node)
           if proper_block_style?(node)
             correct_style_detected
           else
-            add_offense(node, :begin) { opposite_style_detected }
+            add_offense(node, :begin) { unrecognized_style_detected }
```
Comment (jonas054) [inline lib/rubocop/cop/style/block_delimiters.rb]: Actually, those `..._style_detected` methods don't do anything useful when there are more than two styles to choose from. So you can remove this block and the call to `correct_style_detected`.

**Comment by jonas054** [comment]
I just had one more comment. And the Travis build is failing.

**Comment by panthomakos** [comment]
@jonas054 The failures were related to RuboCop rules. I have resolved all of them except for `Class has too many lines. [167/147]` Do you have a recommended way to resolve in this case? I don't see an obvious way to reduce the class size. I'm happy to refactor, but some guidance on direction would be helpful.

**Comment by jonas054** [comment]
How to best refactor could be a long discussion, so for me it's OK if you just run `./bin/rubocop --auto-gen-config` and add/commit the updated `.rubocop_todo.yml`. We can reduce the class length later IMO.

**Comment by panthomakos** [comment]
@jonas054 Ran the `--auto-gen-config` and that has resolved the class length issue.

**Comment by jonas054** [comment]
:+1: Looks good now.

**Comment by bbatsov** [comment]
Drop the `.` from the title of the commit message and rebase on top of the current `master`.

**Comment by panthomakos** [comment]
@bbatsov I always forget the period!

Period removed and branch rebased on master.

**Comment by bbatsov** [comment]
:+1: Thanks!

**Pull request merged**